include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,291 @@
+/* Copyright 2013-2019 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
+ *                     Klaus Steiniger, Felix Schmitt, Benjamin Worpitz,
+ *                     Juncheng E
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include "picongpu/simulation_defines.hpp"
+
+#include <pmacc/dimensions/DataSpaceOperations.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+
+#include <pmacc/mappings/threads/ForEachIdx.hpp>
+#include <pmacc/mappings/threads/IdxConfig.hpp>
+#include <pmacc/memory/Array.hpp>
+#include <pmacc/memory/shared/Allocate.hpp>
+#include <pmacc/mpi/MPIReduce.hpp>
+#include <pmacc/mpi/reduceMethods/Reduce.hpp>
+#include <pmacc/nvidia/atomic.hpp>
+#include <pmacc/nvidia/functors/Add.hpp>
+
+#include "sys/stat.h"
+
+namespace picongpu
+{
+
+/** calculate the scattering of a species
+ *
+ * @tparam T_numWorkers number of workers
+ */
+template <uint32_t T_numWorkers> struct KernelSaxs
+{
+    /**
+     * The SAXS kernel calculates for all particles on the device the
+     * scattering intensity for input calculation ranges.
+     * The parallelization is as follows:
+     *  - The number of threads per block is equal to the number of cells per
+     *    super cells which is also equal to the number of particles per frame
+     *
+     * The procedure starts with calculating unique ids for the threads and
+     * initializing the shared memory.
+     * Then a loop over all super cells starts.
+     * Every thread loads a particle from that super cell and calculates its
+     * scattering structure factor.
+     * For every Particle
+     * exists therefore a unique space within the shared memory.
+     * After that, a thread calculates for a specific scattering vector of all
+     * particles.
```

Comment:
This kernel has quite a lot of parameters. Some can be not obvious for a reader. Please extend this doxygen comment with parameter description (like other functions do it with `@param` in the comments).